src/projection.jl

```diff
@@ -135,13 +132,33 @@ ProjectTo(::Real) = ProjectTo{Real}()
 ProjectTo(::Complex) = ProjectTo{Complex}()
 ProjectTo(::Number) = ProjectTo{Number}()
 for T in (Float16, Float32, Float64, ComplexF16, ComplexF32, ComplexF64)
-    # Preserve low-precision floats as accidental promotion is a common perforance bug
+    # Preserve low-precision floats as accidental promotion is a common perforance bug    
     @eval ProjectTo(::$T) = ProjectTo{$T}()
 end
-ProjectTo(x::Integer) = ProjectTo(float(x))
-ProjectTo(x::Complex{<:Integer}) = ProjectTo(float(x))
-(::ProjectTo{T})(dx::Number) where {T<:Number} = convert(T, dx)
-(::ProjectTo{T})(dx::Number) where {T<:Real} = convert(T, real(dx))
+
+# Preserve low-precision floats as accidental promotion is a common perforance bug
+(::ProjectTo{T})(dx::AbstractFloat) where T<:AbstractFloat = convert(T, dx)
+function (proj::ProjectTo{<:Complex{<:AbstractFloat}})(
+    dx::Union{Complex{<:AbstractFloat}, AbstractFloat}
+)
+    # in this case we can just convert as we know we are dealing with 
+    # boring floating point types or integers
+    return convert(project_type(proj), dx)
+end
```

Comment:
Stylistically, might be tidier to have `where T` and `convert(T, dx)` like the method above, 
```
function (proj::ProjectTo{T})(dx::Union{Complex{<:AbstractFloat}, AbstractFloat}) where {T<:Complex{<:AbstractFloat}}
    # in this case we can just convert as we know we are dealing with 
    # boring floating point types or integers
    return convert(T, dx)
end
```
Or even better to split them:
```suggestion
(::ProjectTo{T})(dx::Complex{<:AbstractFloat}) where {T<:Complex{<:AbstractFloat}} = convert(T, dx)
(::ProjectTo{T})(dx::AbstractFloat) where {T<:Complex{<:AbstractFloat}} = convert(T, dx)
```